### PR TITLE
Stack aware `getFoodComponent`

### PR DESCRIPTION
**Pull request by Phoupraw**
Similar to `getRecipeRemainder`.

### CODE REVIEW

**Review by apple502j — CHANGES_REQUESTED** [review]
Make sure Checkstyle passes before sending a PR...

Also, there are a couple copypastes you might want to fix.

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Please add a test to the testmod.

**Comment by Phoupraw** [comment]
> Please add a test to the testmod.

Added. The test works fine.

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package net.fabricmc.fabric.test.item.gametest;
+
+import net.fabricmc.fabric.api.gametest.v1.FabricGameTest;
+
+public final class FoodGameTest implements FabricGameTest {
```
Comment (modmuss50) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java:37]: Can you finish this?

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package net.fabricmc.fabric.test.item.gametest;
```
Comment (modmuss50) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java:17]: No copyright header

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/CustomDamageTest.java [hunk 3]
```diff
@@ -76,5 +79,13 @@ public ItemStack getRecipeRemainder(ItemStack stack) {
...
 			return ItemStack.EMPTY;
 		}
+
+		@Override
+		public @Nullable FoodComponent getFoodComponent(ItemStack stack) {
+			return new FoodComponent.Builder()
+					.hunger(20 - 20 * stack.getDamage() / stack.getMaxDamage())
+					.saturationModifier(0.5f)
+					.build();
+		}
```
Comment (modmuss50) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/CustomDamageTest.java]: This is in the custom damage test. I think it would be best to move out into its own file.

File: fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/fluid/CombinedProvidersImpl.java [hunk 4]
```diff
@@ -83,8 +83,12 @@ public Storage<FluidVariant> find(ItemStack itemStack, ContainerItemContext cont
 
 	public static Event<FluidStorage.CombinedItemApiProvider> getOrCreateItemEvent(Item item) {
 		// register here is thread-safe, so the query below will return a valid provider (possibly one registered before or from another thread).
-		FluidStorage.ITEM.registerForItems(new Provider(), item);
 		ItemApiLookup.ItemApiProvider<Storage<FluidVariant>, ContainerItemContext> existingProvider = FluidStorage.ITEM.getProvider(item);
+		if (existingProvider == null) {
+			// if you always register, lookup will spam warning, so only register when not registered
+			existingProvider = new Provider();
+			FluidStorage.ITEM.registerForItems(existingProvider, item);
+		}
```
Comment (modmuss50) [inline fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/fluid/CombinedProvidersImpl.java]: Dont make unrelated changes like this, if needed make another PR

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java [hunk 5]
```diff
@@ -100,4 +101,14 @@ public Multimap<EntityAttribute, EntityAttributeModifier> hookGetAttributeModifi
 	public boolean hookIsSuitableFor(Item item, BlockState state) {
 		return item.isSuitableFor((ItemStack) (Object) this, state);
 	}
+
+	/**
+	 * @author Phoupraw
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java]: We generally dont use `@author`, sorry

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java [hunk 6]
```diff
@@ -100,4 +101,14 @@ public Multimap<EntityAttribute, EntityAttributeModifier> hookGetAttributeModifi
...
+	 * @reason use stack-aware {@link #getFoodComponent()} instead
+	 */
+	@Overwrite
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java]: We dont use `@Overwrite` in fabric api

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/FoxEntityMixin.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+abstract class FoxEntityMixin extends AnimalEntity {
+	protected FoxEntityMixin(EntityType<? extends AnimalEntity> entityType, World world) {
+		super(entityType, world);
+	}
+
+	@Redirect(method = {"canEat", "canPickupItem"}, at = @At(value = "INVOKE", target = "Lnet/minecraft/item/Item;isFood()Z",ordinal = 0))
+	private boolean isStackAwareFood(Item instance, ItemStack stack) {
+		return stack.isFood();
+	}
+	@Redirect(method = {"canPickupItem"}, at = @At(value = "INVOKE", target = "Lnet/minecraft/item/Item;isFood()Z",ordinal = 1))
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/FoxEntityMixin.java]: Missing new line, this wont build

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/FoxEntityMixin.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
+package net.fabricmc.fabric.mixin.item;
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/FoxEntityMixin.java:17]: missing copyright header, this wont build.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package net.fabricmc.fabric.mixin.item;
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java:17]: Again, this wont build.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItemStack.java [hunk 10]
```diff
@@ -16,10 +16,13 @@
...
 package net.fabricmc.fabric.api.item.v1;
 
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.item.FoodComponent;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 
-/*
+/**
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItemStack.java]: No need to change this.

**Review by modmuss50 — CHANGES_REQUESTED** [review]

File: fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/fluid/CombinedProvidersImpl.java (hunk 4 above)
Comment (Phoupraw) [reply fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/fluid/CombinedProvidersImpl.java]: I didn't know before that changes from the same branch would be automatically merged into the same PR. My original intent was opening another PR. Now maybe I should roll it back.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java (hunk 6 above)
Comment (Phoupraw) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java]: Always use `@Inject` and `cancel` at `@At("RETURN")`?

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/CustomDamageTest.java (hunk 3 above)
Comment (Phoupraw) [reply fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/CustomDamageTest.java]: I just can't think up other existing NBT for the moment so directly use the most common NBT, damage. Maybe I could use enchantments.

**Comment by Phoupraw** [comment]
wait, the game test doesn't act as I thought.

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+import net.minecraft.util.Identifier;
+import net.minecraft.util.math.Vec3d;
+
+import net.fabricmc.api.ModInitializer;
+import net.fabricmc.fabric.api.gametest.v1.FabricGameTest;
+import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
+
+public final class FoodGameTest implements FabricGameTest, ModInitializer {
+	public static final Item DAMAGE = Registry.register(Registries.ITEM, new Identifier("fabric-item-api-v1-testmod", "damage_food"), new DamageFood(new FabricItemSettings().maxDamage(20)));
+	public static final Item NAME = Registry.register(Registries.ITEM, new Identifier("fabric-item-api-v1-testmod", "name_food"), new NameFood(new FabricItemSettings()));
```
Comment (modmuss50) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java]: I would split the game test and modinitlaiser into two classes.

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+		ItemStack meat = NAME.getDefaultStack();
+		meat.setCustomName(Text.of("mutton"));
+		player.setStackInHand(Hand.MAIN_HAND, meat);
+		player.interact(wolf, Hand.MAIN_HAND);
+		float wolfHealth = wolf.getHealth();
+		context.assertTrue(wolfHealth > 1, "actual %f, expect > 0".formatted(wolfHealth));
+		context.complete();
+	}
+
+	//@GameTest(templateName = EMPTY_STRUCTURE)
```
Comment (modmuss50) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java]: Commented out code.

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/CustomDamageTest.java (hunk 3 above)
Comment (modmuss50) [reply fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/CustomDamageTest.java]: Ok, thats fine maybe just put a comment on the method saying its for the stackaware food test.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import net.minecraft.item.FoodComponent;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.Hand;
+
+@Mixin(WolfEntity.class)
+class WolfEntityMixin {
+	@Unique
+	private ItemStack fabric_interactMob_itemStack = ItemStack.EMPTY;
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java]: Should be a thread local

Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java]: No need to prefix it with `fabric_` when using `@Unique`

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+	@Inject(method = "interactMob", at = @At("HEAD"))
+	private void storeCopy(PlayerEntity player, Hand hand, CallbackInfoReturnable<ActionResult> cir) {
+		fabric_interactMob_itemStack = player.getStackInHand(hand).copy();
+	}
+
+	@Redirect(method = "interactMob", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/Item;getFoodComponent()Lnet/minecraft/item/FoodComponent;"))
+	private @Nullable FoodComponent getStackAwareFoodComponent(Item instance, PlayerEntity player, Hand hand) {
+		return fabric_interactMob_itemStack.getFoodComponent();
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java]: Reset the temporary itemstack here once done with it.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItem.java [hunk 15]
```diff
@@ -122,4 +124,15 @@ default boolean isSuitableFor(ItemStack stack, BlockState state) {
 	default ItemStack getRecipeRemainder(ItemStack stack) {
 		return ((Item) this).hasRecipeRemainder() ? ((Item) this).getRecipeRemainder().getDefaultStack() : ItemStack.EMPTY;
 	}
+
+	/**
+	 * This is a stack-aware version of {@link Item#getFoodComponent()}.
+	 * Note that simple item remainders can also be set via {@link Item.Settings#food(FoodComponent)}.
+	 * If you want to get a remainder for a stack, is <b>recommended</b> to use the stack version of this method: {@link FabricItemStack#getFoodComponent()}.
```
Comment (Shnupbups) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItem.java]: Javadoc still references remainders instead of food components

**Review by Shnupbups — CHANGES_REQUESTED** [review]

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItemStack.java [hunk 16]
```diff
@@ -36,4 +39,14 @@ public interface FabricItemStack {
 	default ItemStack getRecipeRemainder() {
 		return ((ItemStack) this).getItem().getRecipeRemainder((ItemStack) this);
 	}
+
+	/**
+	 * Stack-aware version of {@link Item#getFoodComponent()}.
+	 * See {@link FabricItem#getFoodComponent(ItemStack)} for a more in depth description.
+	 *
+	 * @return this itemStack's {@link FoodComponent}, or {@code null} if none was set.
```
Comment (Juuxel) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItemStack.java]:
```suggestion
	 * @return this item stack's {@link FoodComponent}, or {@code null} if none was set
```

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItem.java [hunk 17]
```diff
@@ -122,4 +124,15 @@ default boolean isSuitableFor(ItemStack stack, BlockState state) {
 	default ItemStack getRecipeRemainder(ItemStack stack) {
 		return ((Item) this).hasRecipeRemainder() ? ((Item) this).getRecipeRemainder().getDefaultStack() : ItemStack.EMPTY;
 	}
+
+	/**
+	 * This is a stack-aware version of {@link Item#getFoodComponent()}.
+	 * Note that simple food component can also be set via {@link Item.Settings#food(FoodComponent)}.
+	 * If you want to get a food component for a stack, is <b>recommended</b> to use the stack version of this method: {@link FabricItemStack#getFoodComponent()}.
+	 *
+	 * @return this item's {@link FoodComponent}, or {@code null} if none was set.
```
Comment (Juuxel) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItem.java]:
```suggestion
	 * @return this item's {@link FoodComponent}, or {@code null} if none was set
```

**Review by Juuxel — COMMENTED** [review]
Some notes about the javadoc

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Redirect;
+
+import net.minecraft.item.BlockItem;
+import net.minecraft.item.ItemUsageContext;
+
+@Mixin(BlockItem.class)
+class BlockItemMixin {
+	@Redirect(method = "useOnBlock", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/BlockItem;isFood()Z"))
```
Comment (modmuss50) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java:28]: Could a lot of these mixins be improved by using WrapOperation?

**Review by modmuss50 — APPROVED** [review]
This PR looks good, my main worry is the amount of Mixins and @Redirect's. I guess there isnt any other way around it. Might be nice to explore using Mixin Extras now that we have it.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java (hunk 18 above)
Comment (Phoupraw) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java:28]: At that moment IJ Idea and Minecraft Dev plugin doesn't support mixin extra well, so I use redirect instead to avoid potential a mass of errors.

Comment (LlamaLad7) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java:28]: MCDev has had full mixinextras support for a long time now.

Comment (Phoupraw) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java:28]: Ah? I had a try just now, and found that it still suggests incorrect return type wrapping putting field.
For example, for the following wrap:
```java
@WrapOperation(method = "<clinit>", at = @At(value = "FIELD", target = "Lnet/fabricmc/fabric/api/transfer/v1/fluid/FluidStorage;ITEM:Lnet/fabricmc/fabric/api/lookup/v1/item/ItemApiLookup;", opcode = Opcodes.PUTSTATIC))
```
the correct handler method signature should be
```java
private static void customField(ItemApiLookup<Storage<FluidVariant>, ContainerItemContext> value, Operation<ItemApiLookup<Storage<FluidVariant>, ContainerItemContext>> original)
```
but minecraft dev plugin suggests that it should be
```java
private static ItemApiLookup<Storage<FluidVariant>, ContainerItemContext> customField(ItemApiLookup<Storage<FluidVariant>, ContainerItemContext> value, Operation<ItemApiLookup<Storage<FluidVariant>, ContainerItemContext>> original)
```
The latter causes an error on game launching.

Comment (LlamaLad7) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java:28]: Ok you might be right about that one… will fix
Don’t know of any other such cases

Comment (Phoupraw) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java:28]: I just found that fabric api project still remains loader 0.14.21... And fabric item api still depends on fabricloader >=0.4.0 in fabric.mod.json... Shall we update them?

Comment (modmuss50) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BlockItemMixin.java:28]: We had the same problem in https://github.com/FabricMC/fabric/pull/3403#discussion_r1437078861 and decided on using the @Redirect's in 1.20.1/.2 and then for .4 using ME. Quite happy to leave this as-is and when cherry-picking to 1.20.4 I can change this to use ME.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItem.java [hunk 19]
```diff
@@ -122,4 +124,15 @@ default boolean isSuitableFor(ItemStack stack, BlockState state) {
 	default ItemStack getRecipeRemainder(ItemStack stack) {
 		return ((Item) this).hasRecipeRemainder() ? ((Item) this).getRecipeRemainder().getDefaultStack() : ItemStack.EMPTY;
 	}
+
+	/**
+	 * This is a stack-aware version of {@link Item#getFoodComponent()}.
+	 * Note that simple food component can also be set via {@link Item.Settings#food(FoodComponent)}.
+	 * If you want to get a food component for a stack, is <b>recommended</b> to use the stack version of this method: {@link FabricItemStack#getFoodComponent()}.
```
Comment (Juuxel) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItem.java]:
```suggestion
	 * If you want to get a food component for a stack, is <strong>recommended</strong> to use the stack version of this method: {@link FabricItemStack#getFoodComponent()}.
```
Using `<strong>` for this use case is a [common HTML recommendation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/b).

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+import net.minecraft.item.FoodComponent;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.Hand;
+
+@Mixin(WolfEntity.class)
+class WolfEntityMixin {
+	@SuppressWarnings("MissingUnique")
+	private static final ThreadLocal<ItemStack> fabric_interactMob_itemStack = new ThreadLocal<>();
```
Comment (Juuxel) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java]: Nit: static final fields should use CONSTANT_CASE

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+		WolfEntity wolf = context.spawnEntity(EntityType.WOLF, context.getRelative(Vec3d.ZERO));
+		wolf.setTamed(true);
+		wolf.setOwner(player);
+		wolf.setHealth(1f);
+		ItemStack meat = FoodGameInitializer.NAME.getDefaultStack();
+		meat.setCustomName(Text.of("mutton"));
+		player.setStackInHand(Hand.MAIN_HAND, meat);
+		player.interact(wolf, Hand.MAIN_HAND);
+		float wolfHealth = wolf.getHealth();
+		context.assertTrue(wolfHealth > 1, "actual %f, expect > 0".formatted(wolfHealth));
```
Comment (Juuxel) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/gametest/FoodGameTest.java]: The message doesn't match the check (`> 0` vs `> 1`)

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java (hunk 20 above)
Comment (Phoupraw) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java]: itemStack is the name of the local variable. interactMob is the method name where the var is. They should be kept. So the prefix, fabric, should be FABRIC?

Comment (Juuxel) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java]: - Vanilla local variables do not have Yarn names, this is just something generated by Tiny Remapper
- It does not need a Fabric prefix, `@Unique` is enough
- It also doesn't need the exact method name.

Something like `@Unique ThreadLocal<ItemStack> INTERACTION_STACK` works well. See another similar example from this module:
https://github.com/FabricMC/fabric/blob/1681346ec99b3030403d3eedd2773bf2d8d38c4e/fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/AbstractFurnaceBlockEntityMixin.java#L36-L37

Comment (modmuss50) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/WolfEntityMixin.java]: I have fixed this.

**Comment by modmuss50** [comment]
I plan to do another release very shortly, I will re-target this PR to 1.20.4 and use ME.

**Comment by modmuss50** [comment]
I have created a 1.20.4 + ME version of this PR here: https://github.com/FabricMC/fabric/pull/3520

**Comment by modmuss50** [comment]
Merged into 1.20.4 via https://github.com/FabricMC/fabric/pull/3520 Many thanks for this 👍
